tests/unittests/MemoryAllocatorTest.cpp

```diff
@@ -342,3 +342,9199 @@ TEST(MemAlloc, testAlignment) {
+                            "Allocations are invalid!");
+#endif
+}
+
+/// Test invalid FREE before ALLOC for allocateAll.
+TEST(MemAlloc, testAllocateAllInvalidFreeBeforeAlloc) {
+  MemoryAllocator MA("test", 0, 64);
+  void *handle = reinterpret_cast<void *>(0);
+  std::list<Allocation> allocList;
+  allocList.push_back(Allocation(handle, false, 0));
+  allocList.push_back(Allocation(handle, true, 10));
+#ifndef NDEBUG
+  ASSERT_DEATH_IF_SUPPORTED(MA.allocateAll(allocList),
+                            "Allocations are invalid!");
+#endif
+}
+
+/// Test invalid handles for allocateAll.
+TEST(MemAlloc, testAllocateAllInvalidHandles1) {
+  MemoryAllocator MA("test", 0, 64);
+  void *handle = reinterpret_cast<void *>(0);
+  std::list<Allocation> allocList;
+  allocList.push_back(Allocation(handle, true, 10));
+  allocList.push_back(Allocation(handle, true, 10));
+  allocList.push_back(Allocation(handle, false, 0));
+  allocList.push_back(Allocation(handle, false, 0));
+#ifndef NDEBUG
+  ASSERT_DEATH_IF_SUPPORTED(MA.allocateAll(allocList),
+                            "Allocations are invalid!");
+#endif
+}
+
+/// Test invalid handles for allocateAll.
+TEST(MemAlloc, testAllocateAllInvalidHandles2) {
+  MemoryAllocator MA("test", 0, 64);
+  void *handle0 = reinterpret_cast<void *>(0);
+  void *handle1 = reinterpret_cast<void *>(0);
+  std::list<Allocation> allocList;
+  allocList.push_back(Allocation(handle0, true, 10));
+  allocList.push_back(Allocation(handle0, true, 10));
+  allocList.push_back(Allocation(handle1, false, 0));
+  allocList.push_back(Allocation(handle1, false, 0));
+#ifndef NDEBUG
+  ASSERT_DEATH_IF_SUPPORTED(MA.allocateAll(allocList),
+                            "Allocations are invalid!");
+#endif
+}
+
+/// Test segment size 0 for allocateAll.
+TEST(MemAlloc, testAllocateAllSize0) {
+  MemoryAllocator MA("test", 0, 64);
+  void *handle = reinterpret_cast<void *>(0);
+  std::list<Allocation> allocList;
+  allocList.push_back(Allocation(handle, true, 0));
+  allocList.push_back(Allocation(handle, false, 0));
+  uint64_t usedSize = MA.allocateAll(allocList);
+  EXPECT_EQ(usedSize, 0);
+  EXPECT_EQ(MA.getSize(handle), 0);
+}
+
+/// Test empty allocs for allocateAll.
+TEST(MemAlloc, testAllocateAllEmptyAlloc) {
+  MemoryAllocator MA("test", 0, 64);
+  std::list<Allocation> allocList;
+  uint64_t usedSize = MA.allocateAll(allocList);
+  EXPECT_EQ(usedSize, 0);
+}
+
+/// Test memory overflow for allocateAll.
+TEST(MemAlloc, testAllocateAllMemOverflow) {
+  MemoryAllocator MA("test", 10, 64);
+  void *handle = reinterpret_cast<void *>(0);
+  std::list<Allocation> allocList;
+  allocList.push_back(Allocation(handle, true, 100));
+  allocList.push_back(Allocation(handle, false, 0));
+  uint64_t usedSize = MA.allocateAll(allocList);
+  EXPECT_EQ(usedSize, MemoryAllocator::npos);
+#ifndef NDEBUG
+  ASSERT_DEATH_IF_SUPPORTED(MA.getSize(handle), "Unknown handle");
+#endif
+}
+
+/// Test memory allocation for model cifar10_quant.tflite.
```

Comment:
If I understand this code correctly you recorded the memory allocation requests for a number of models and try to check how they are handled. I'm a bit concerned that we hard-code those numbers here. Loading a model and doing memory allocation would be more reliable....

Reply:
Also, most of the code in this file is auto-generated I assume. And the file is huge. Could you instead dump the allocs per model into an input file and then in this CPP file just read them, create a sequence of allocations and use the allocator then? This would make this source file MUCH smaller.

Reply:
Indeed I recorded the allocations from 79 models (by parsing their corresponding IR) and auto-generated unit tests here using a Matlab script. I don't see why hardcoding would be a problem since the unit tests implicitly verify the allocation is correct (segments are always validated within the function `allocateAll`) and also have a regression of the allocation efficiency (you can verify that for each unit tests the efficiency is checked against the reference) which is more than enough to test the allocator.
Compiling all the models directly would be an overkill since that means all the models should be downloaded during testing (this means 2.36 GB).
I agree we could compress the unit tests. We have 2 options:
1) either keep only a couple of unit tests for models (say 10)
2) we could export the information into files but in order to save space we should export them as binaries but this makes the tests harder to read and requires some extra code for parsing. Any suggestions?

Reply:
Yes, compiling all models would be an overkill. Let's compress the unit tests.

> we could export the information into files but in order to save space we should export them as binaries but this makes the tests harder to read and requires some extra code for parsing. Any suggestions?

You could use text files with the recorded allocs. The idea is to keep this information that is purely data and semantically is not very important out of sight when you look at the test. The test should be small and simple:
- load data from a file
- build a list of allocs
- run the allocator
- check the results/quality of allocations

Reply:
Should I make one text file for each model (e.g. `Model1.txt`, `Model2.txt`, etc)? Where to put these files: in a separate folder inside the unit tests is ok, e.g. `MemoryAllocatorTestModels`?
Or a single file called `MemoryAllocatorTest.txt` in the same folder?

Reply:
> in a separate folder inside the unit tests is ok, e.g. MemoryAllocatorTestModels?

Yes, this would be the better solution.

Reply:
This is done!